node/src/reactor/validator/tests.rs

```diff
@@ -241,23 +214,29 @@ async fn run_equivocator_network() {
         .await
         .expect("network initialization failed");
 
+    // With an auction delay of 1 and an unbonding delay of 3, there are always 2 bonded past
+    // eras. The era supervisor keeps five active eras in memory — the oldest two without units,
+    // just for validating evidence if needed. So if we run this for five eras, the test can catch
+    // issues with unbonding and dropping obsolete eras: In era 5, the units from era 2 are
+    // dropped, and era 0 is dropped completely.
```

Comment:
Can we extract the calculation of (based on the `unbonding_delay` and `bonding_delay`):
* how many eras are bonded,
* which eras are "evidence only"
* which eras are being dropped

to a method and call it in both Highway and here? So that it's always in line?

Reply:
I'm not really sure what to extract:
* [This subtraction](https://github.com/CasperLabs/casper-node/blob/cb1d20ad1ea6e245cd8237f9406885a1e785c669/node/src/components/consensus/era_supervisor.rs#L166) determines `B`.
* [Here](https://github.com/CasperLabs/casper-node/blob/cb1d20ad1ea6e245cd8237f9406885a1e785c669/node/src/components/consensus/era_supervisor.rs#L364) we drop an old era.
* [And here](https://github.com/CasperLabs/casper-node/blob/cb1d20ad1ea6e245cd8237f9406885a1e785c669/node/src/components/consensus/era_supervisor.rs#L371) we make it "evidence only".

What method would be be calling here in the test?
Do you mean something like asserting that `5 - 2 B - 1 ≥ 0`?

Reply:
```rust
fn bonded_eras(&HigwayProtocol) -> u64 { … }

fn obsolete_era_id(curr_era: EraId) -> Option<u64> { … }

fn evidence_only_era_id(curr_era: EraId) -> Option<u64> { … }
```

make them a "stand-only" methods (not attached to the instance of Highway or anything) and call them in the test when verifying whether we should have already removed era 0.

Reply:
This has nothing to do with Highway; it's consensus-independent era supervisor logic.

I agree we should make `bonded_eras` a method, and actually remove the field from `EraSupervisor`. It could be a method of `components::consensus::config::ProtocolConfig`?

And `obsolete_era_id` and `evidence_only_era_id` would also be methods of `ProtocolConfig`?

The only reason why I'm hesitating is: I'm not sure we should have any logic at all in a _config_ impl?

Reply:
It doesn't have to be a method in the _config_ impl. It can be a standalone method in a file, right? Just like we have methods in `state.rs`.

Reply:
> This has nothing to do with Highway; it's consensus-independent era supervisor logic.

Right, I saw `protocol_config.auction_delay` and I thought it's a highway config.

Reply:
OK, I could add the three as stand-alone functions to `era_supervisor.rs`. Not sure whether their argument should bethe protocol config, or `auction_delay` and `unbonding_delay`.

Reply:
I'll make it the protocol config: That's available in both the test and the era supervisor.

Reply:
Removing the field (https://github.com/CasperLabs/casper-node/pull/1052/commits/63a5611572d606bb08a35144c294078eeb867203) makes a lot of sense to me.

I added an `oldest_bonded_era` function, but not one for the oldest evidence-only era in https://github.com/CasperLabs/casper-node/pull/1052/commits/be81f09c823446df700daaa3b87b1cfee30ac443 — not really sure whether that commit makes things better and whether it's at all what you have in mind?

Also note that some of this logic will change with CEP 36.

Reply:
I'm wondering whether it would also make sense to extract these two:
```rust
Some(obsolete_era_id) = era_id.checked_sub(2 * self.bonded_eras() + 1)
```
and
```rust
Some(evidence_only_era_id) = era_id.checked_sub(self.bonded_eras() + 1)
```
? It is an implementation detail we kind of assume in other places – especially in this test right now.

Reply:
I did, more or less, in the last commit.
I extracted only one method, though, and without the final `1`, because I believe "the oldest bonded era" is more meaningful than "the youngest not-bonded-anymore era".